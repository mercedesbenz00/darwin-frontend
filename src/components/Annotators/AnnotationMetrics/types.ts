import { MembershipPayload, DatasetAnnotationReportPayload } from '@/store/types'

export type AnnotationDataRange = 'total' | 'month' | 'week' | 'day'
export type AnnotationDataGranularity = 'day' | 'hour'

export type ChartDimension =
  'annotationTime' |
  'imagesAnnotated' |
  'avgTimePerAnnotation' |
  'annotationsCreated' |
  'reviewPassRate'

export type AdaptedDataPoint =
  {[k in Exclude<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number } &
  {[k in Extract<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number | null} &
  { userId: number | null, date: string }

export type AggregatedDataPoint = Pick<
    DatasetAnnotationReportPayload,
    | 'annotation_time'
    | 'annotations_approved'
    | 'annotations_created'
    | 'date'
    | 'images_annotated'
    | 'images_approved'
    | 'images_rejected'
    | 'user_id'
  >

export type AnnotationDimensionData = {
  annotationTime: number,
  annotationsCreated: number,
  avgTimePerAnnotation: number | null,
  imagesAnnotated: number,
  reviewPassRate: number | null
}

export type AnnotationChartData = {
  data: AdaptedDataPoint[]
  dimensions: AnnotationDimensionData
  member?: MembershipPayload
  visible: boolean
}

export type AnnotationBarChartData = {
  data: AggregatedDataPoint[]
  member?: MembershipPayload
}

/**
 * Format for a single item to be shown in a chart tooltip.
 *
 * A tooltip will contain multiple items
 */
export type ChartTooltipData = { member: MembershipPayload | null, value: string}
