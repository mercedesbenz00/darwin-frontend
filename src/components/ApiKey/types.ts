export type AbilityGroup =
  'manage_comments' |
  'manage_tasks' |
  'manage_datasets' |
  'manage_team' |
  'manage_team_members' |
  'manage_models'

export type Group = { id: AbilityGroup; name: string }
export type GroupOption = Group & { selected: boolean }

export enum SupportedLanguages {
  shell = 'shell',
  python = 'python',
  javascript = 'javascript',
  elixir = 'elixir',
  cli = 'cli'
}

export type AccessCodeExample = { language: SupportedLanguages, code: string }
