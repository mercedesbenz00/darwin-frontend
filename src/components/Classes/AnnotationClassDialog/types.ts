import { AnnotationClassPayload, AnnotationClassMetadata, AnnotationTypeName } from '@/store/types'

type PendingFields =
  | 'crop_key'
  | 'crop_url'
  | 'image_height'
  | 'image_width'
  | 'key'
  | 'original_image_url'
  | 'scale'
  | 'x'
  | 'y'

/**
 * Holds a class image that isn't necessarily persisted to the backend yet.
 * That means it might be missing any of the url fields
 *
 * It can be saved to backend once at least the `key` field is set, as that one
 * can be used to infer/compute all the other fields.
 *
 * - `crop_key` - repleace "original" with "crop" in `key` field
 * - `crop_url` - signed s3 url (by backend), for `crop_key`
 * - `original_image_url` - signed s3 url (by backend) for `key`
 */
export type PendingClassImage =
  Omit<AnnotationClassPayload['images'][0], PendingFields>
  & Partial<Pick<AnnotationClassPayload['images'][0], PendingFields>>

// eslint-disable-next-line camelcase
export type PendingClassImageWithUrlSet = PendingClassImage & { original_image_url: string }

export const isClassImage = (
  image: PendingClassImage
): image is AnnotationClassPayload['images'][0] =>
  !!image.key && !!image.crop_key

export type AnnotationClassData = {
  datasets: AnnotationClassPayload['datasets']
  description: string | null
  hotkey: null | string
  id: number | null
  images: PendingClassImage[]
  metadata: AnnotationClassMetadata
  name: string | null
}

export type DefaultAnnotationClassData = Partial<AnnotationClassData> & {
  selectedMainAnnotationType?: AnnotationTypeName
}

export type AnnotationClassValidationErrors = {
  annotationTypes?: string
  name?: string
  skeleton?: string
}

export const VIEWPORT_WIDTH: number = 150
export const VIEWPORT_HEIGHT: number = 150

export type CropInfo = {
  orientation: number
  points: [string, string, string, string]
  zoom: number
}

/**
 * VueCroppie is really archaic and requires us to get crop info via $ref,
 * so this type defines that ref
 */
export type Croppie = {
  /** Binds options to the underlying croppie library */
  bind: (options: Pick<CropInfo, 'points' | 'zoom'> & { url: string}) => void

  /** Retrieves current crop info */
  get: () => CropInfo
  result: (options?: Record<string, string>, callback?: Function) => void

  /** Sets scale/zoom level of cropper */
  setZoom: (value: number) => void
}
