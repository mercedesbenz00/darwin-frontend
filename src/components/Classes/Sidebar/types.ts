import { AnnotationTypeName } from '@/store/types'

export type AnnotationTypeFilterItemType = {
  id: string | number
  label: string
  icon?: string
  name: AnnotationTypeName
  color: string
  count: number | null
  [key: string]: any
}
