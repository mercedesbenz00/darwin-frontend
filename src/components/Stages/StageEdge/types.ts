import { StageType } from '@/store/types/StageType'
import { V2WorkflowEdgePayload } from '@/store/types/V2WorkflowEdgePayload'
import { StageCoords } from '@/store/types/V2WorkflowStagePayload'

export enum EdgeType {
  IN = 'in',
  OUT = 'out'
}

export type StageEdgeProps = {
  id: string
  edge: V2WorkflowEdgePayload | null
  edgeType: EdgeType
  connectable: boolean
  type: StageType
  currentStagePosition: StageCoords
  name?: string
}

export type MouseClickPosition = {
  mouseX: number
  mouseY: number
}
