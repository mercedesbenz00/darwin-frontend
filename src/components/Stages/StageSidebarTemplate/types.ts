import { StageType } from '@/store/types/StageType'
import { StageCoords } from '@/store/types/V2WorkflowStagePayload'

export type StageSidebarTemplateProps = {
  type: StageType
}

export type DragEnd = {
  type: StageType
  droppable: boolean
} & StageCoords
