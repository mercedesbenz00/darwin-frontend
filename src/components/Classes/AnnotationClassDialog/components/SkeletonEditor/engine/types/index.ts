export * from './SkeletonEdgeType'
export * from './SkeletonNodeType'
