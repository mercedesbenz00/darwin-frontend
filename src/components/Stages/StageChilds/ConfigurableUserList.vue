<template>
  <div>
    <div class="sul-input__wrapper">
      <input-field
        placeholder="Search Users"
        :value="searchValue"
        @change="setSearch"
      >
        <template #left-icon>
          <icon-mono-search />
        </template>
      </input-field>
    </div>
    <div class="sul-scrollable__wrapper">
      <div class="sul-content__wrapper">
        <anyone-list-element
          @click.native="selectAnyone"
          :selected="assignableTo === 'anyone'"
        />
      </div>
      <div
        class="sul-content__wrapper"
      >
        <user-list-element
          v-for="member in searchResults"
          :key="member.id"
          :member="member"
          :selected="isMemberSelected(member.user_id)"
          @click.native="selectMember(member)"
        />
      </div>
      <div v-if="searchValue.length > 0 && searchResults.length === 0">
        <h3 class="sul-result__label">
          No user found called â€ž{{ searchValue }}".
        </h3>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref, watch } from 'vue'

import { IconMonoSearch } from '@/assets/icons/V2/Mono'
import { InputField } from '@/components/Common/InputField/V2'
import AnyoneListElement from '@/components/Stages/StageChilds/AnyoneListElement.vue'
import UserListElement from '@/components/Stages/StageChilds/UserListElement.vue'
import { searchByNameV2 } from '@/components/Stages/utils'
import { useTeamStore } from '@/composables/useTeamStore'
import { useEditedWorkflow } from '@/pinia/useEditedWorkflow'
import { MembershipPayload } from '@/store/types/MembershipPayload'
import { V2AnnotateStagePayload, V2ReviewStagePayload } from '@/store/types/V2WorkflowStagePayload'

export default defineComponent({
  name: 'ConfigurableUserList',
  components: {
    AnyoneListElement,
    IconMonoSearch,
    InputField,
    UserListElement,
  },
  props: {
    stage: {
      type: Object as () => V2AnnotateStagePayload | V2ReviewStagePayload,
      required: true
    }
  },
  setup (props) {
    const assignableTo = ref<typeof props.stage['config']['assignable_to']>('anyone')
    const assigneeIds = ref<number[]>([])

    const { updateStageConfig, updateAssignableUsers } = useEditedWorkflow()

    const setData = (): void => {
      const assignableIds = (props.stage.assignable_users || []).map(a => a.user_id)
      assignableTo.value = props.stage.config.assignable_to
      assigneeIds.value = assignableIds
    }

    watch(() => props.stage, () => setData(), { immediate: true })

    const isMemberSelected = (id: number): boolean =>
      assigneeIds.value.some(userId => userId === id)

    const submitData = (): void => {
      updateStageConfig({
        stageId: props.stage.id,
        config: {
          assignable_to: assignableTo.value
        }
      })

      updateAssignableUsers({
        stageId: props.stage.id,
        assignable_users: assigneeIds.value.map(id => ({  user_id: id, stage_id: props.stage.id }))
      })
    }

    const selectAnyone = (): void => {
      assignableTo.value = 'anyone'
      assigneeIds.value = []
      submitData()
    }

    const selectMember = (member: MembershipPayload): void => {
      const index = assigneeIds.value.findIndex(id => id === member.user_id)

      if (index > -1) {
        assigneeIds.value.splice(index, 1)
      } else {
        assigneeIds.value.push(member.user_id)
      }

      assignableTo.value = (assigneeIds.value.length > 0) ? 'manual' : 'anyone'

      submitData()
    }

    const searchValue = ref('')

    const setSearch = (val: string): void => {
      searchValue.value = val
    }

    const { allRelevantMemberships } = useTeamStore()

    const searchResults = computed<MembershipPayload[]>(
      () => searchValue.value.length > 0
        ? searchByNameV2(allRelevantMemberships.value, searchValue.value)
        : allRelevantMemberships.value
    )

    return {
      assignableTo,
      isMemberSelected,
      searchResults,
      searchValue,
      selectAnyone,
      selectMember,
      setSearch
    }
  }
})
</script>

<style lang="scss" scoped>
.sul-input__wrapper {
  display: block;
  padding: 8px;
}

.sul-scrollable__wrapper {
  display: block;
  height: auto;
  overflow: scroll;
  padding: 0 8px 8px 8px;

  @include hidden-scrollbar;
}

.sul-content__wrapper {
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 1px;
  padding: 1.5px 0;

  &:last-child {
    border-top: 1px solid $colorNeutralsLight300;
  }
}

.sul-result__label {
  @include typography(md-1, inter, 500);
  color: $colorNeutralsLight500;
  text-align: center;
  padding: 6px 0;
}
</style>
