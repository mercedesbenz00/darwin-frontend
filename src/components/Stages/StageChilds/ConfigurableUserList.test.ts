import { createLocalVue, shallowMount } from '@vue/test-utils'
import VTooltip from 'v-tooltip'
import Vuex from 'vuex'

import { stubDirectiveWithAttribute } from 'test/unit/directiveStubs'
import {
  buildMembershipPayload,
  buildTeamPayload,
  buildV2WorkflowPayload,
  buildV2WorkflowStagePayload
} from 'test/unit/factories'

import ConfigurableUserList from '@/components/Stages/StageChilds/ConfigurableUserList.vue'
import V2Workflow, { getInitialState as V2WorkflowState } from '@/store/modules/V2Workflow'
import team, { getInitialState as teamState } from '@/store/modules/team'
import workview, { getInitialState as workviewState } from '@/store/modules/workview'
import { StageType } from '@/store/types/StageType'
import { V2AnnotateStagePayload, V2ReviewStagePayload } from '@/store/types/V2WorkflowStagePayload'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VTooltip)
localVue.directive('tooltip', stubDirectiveWithAttribute)

const workflow = buildV2WorkflowPayload({
  stages: [buildV2WorkflowStagePayload({ type: StageType.Review })]
})

const v7 = buildTeamPayload({ id: 1 })

let propsData: { stage: V2AnnotateStagePayload | V2ReviewStagePayload; }

const store = new Vuex.Store({
  modules: {
    team: { ...team, state: teamState() },
    workview: { ...workview, state: workviewState() },
    v2Workflow: { ...V2Workflow, state: V2WorkflowState() }
  }
})

beforeEach(() => {
  jest.spyOn(store, 'dispatch').mockResolvedValue({})

  store.commit('team/SET_CURRENT_TEAM', v7)

  const membershipList = [
    buildMembershipPayload({
      id: 1,
      user_id: 1,
      first_name: 'Jack',
      last_name: 'Annotator ',
      team_id: v7.id
    }),
    buildMembershipPayload({
      id: 2,
      user_id: 2,
      first_name: 'Sam',
      last_name: 'Annotator ',
      team_id: v7.id
    })
  ]

  store.commit('team/SET_MEMBERSHIPS', membershipList)
  propsData = {
    stage: {
      ...workflow.stages[0],
      assignable_users: [{ user_id: 1 }, { user_id: 2 }],
    } as V2ReviewStagePayload
  }

})

it('should match snapshot', () => {
  const wrapper = shallowMount(ConfigurableUserList, {
    propsData,
    localVue,
    store
  })
  expect(wrapper).toMatchSnapshot()
})
