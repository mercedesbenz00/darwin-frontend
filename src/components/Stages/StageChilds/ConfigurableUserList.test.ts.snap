// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div class="sul-input__wrapper">
    <input-field-stub value="" tag="div" placeholder="Search Users" type="text" items=""></input-field-stub>
  </div>
  <div class="sul-scrollable__wrapper">
    <div class="sul-content__wrapper">
      <anyone-list-element-stub></anyone-list-element-stub>
    </div>
    <div class="sul-content__wrapper">
      <user-list-element-stub member="[object Object]" selected="true"></user-list-element-stub>
      <user-list-element-stub member="[object Object]" selected="true"></user-list-element-stub>
    </div>
    <!---->
  </div>
</div>
`;
